src/lib/gadgets/bitwise.ts

```diff
@@ -0,0 +1,130 @@
+import { Provable } from '../provable.js';
+import { Field as Fp } from '../../provable/field-bigint.js';
+import { Field } from '../field.js';
+import * as Gates from '../gates.js';
+
+export { xor };
+
+function xor(a: Field, b: Field, length: number) {
+  // check that both input lengths are positive
+  assert(length > 0, `Input lengths need to be positive values.`);
+
+  // check that length does not exceed maximum field size in bits
+  assert(
+    length <= Field.sizeInBits(),
+    `Length ${length} exceeds maximum of ${Field.sizeInBits()} bits.`
+  );
+
+  // handle constant case
+  if (a.isConstant() && b.isConstant()) {
+    assert(
+      a.toBigInt() < 2 ** length,
+      `${a.toBigInt()} does not fit into ${length} bits`
+    );
+
+    assert(
+      b.toBigInt() < 2 ** length,
+      `${b.toBigInt()} does not fit into ${length} bits`
+    );
```

Comment:
To match the behaviour of the provable version, `2 ** length` should be replace by `2 ** padLength`. Also, it should be done with bigints to get an exact result. To make this work you'll need to move the definition of `padLength` to above this.
```suggestion
   let max = 1n << BigInt(padLength);
    assert(
      a.toBigInt() < max,
      `${a.toBigInt()} does not fit into ${padLength} bits`
    );

    assert(
      b.toBigInt() < max,
      `${b.toBigInt()} does not fit into ${padLength} bits`
    );
```